tools/c7n_openstack/c7n_openstack/resources/server.py

```diff
@@ -231,3 +230,84 @@ def find_object_by_property(collection, k, v):
         return None
     assert len(result) == 1
     return result[0]
+
+
+@Server.filter_registry.register('security-group')
+class SecurityGroupFilter(ListItemFilter):
+    """Filters Servers based on attached security groups attributes
+
+    :example:
+
+    Finds servers with a security group that is stateful
+    .. code-block:: yaml
+
+            policies:
+              - name: server-securitygroup-stateful
+                resource: openstack.server
+                filters:
+                  - type: security-group
+                    attrs:
+                      - type: value
+                        key: stateful
+                        value: true
+
+    :example:
+
+    Finds servers with a security group that has inbound rules which
+    include 0.0.0.0/0 and any port
+
+    .. code-block:: yaml
+
+            policies:
+              - name: server-securitygroup-open-to-internet
+                resource: openstack.server
+                filters:
+                  - type: security-group
+                    key: security_group_rules
+                    attrs:
+                      - type: value
+                        key: direction
+                        value: ingress
+                      - type: value
+                        key: !port_range_min && !port_range_max
+                        value: true
+                      - type: value
+                        key: remote_ip_prefix
+                        value: '0.0.0.0/0'
+    """
+    schema = type_schema(
+        'security-group',
+        key= {'type': 'string'},
+        attrs={'$ref': '#/definitions/filters_common/list_item_attrs'}
+    )
+
+    annotate_items = True
+    _sg_map = dict()
```

Comment:
fwiw, its generally best to use an instance variable versus a mutable class variable.